### PR TITLE
ts 1.0

### PR DESCRIPTION
**Pull request by DanielDavidyan**

### CODE REVIEW

File: src/ts/Product.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        this.products = []
+        this.noOfProducts = 0;
+        this.totalPrice = 0;
+    }
+}
+
+(function main() {
+    console.log("---------------------------------------------------------");
+
+    let prod1 = new Product("bamba", "nice", 5, "www.bamba.co.il", 20);
```
[the remark this answers is not in the document]
Comment (1Like1Prayer) [reply src/ts/Product.ts:58]: why let?

File: src/ts/Product.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    let prod2 = new Product("bisli", "great", 10, "www.bisli.co.il", 10);
+    let prod3 = new Product("apropo", "good", 2, "www.apropo.co.il", 5);
+
+    console.log("check the method updateProductAmount");
+    console.log("the limit of " + prod1.name + " before update is = " + prod1.limit);
+    prod1.updateProductAmount(50);
+    console.log("the limit of " + prod1.name + " after update is = " + prod1.limit);
+
+    let cart1 = new Cart();
```
Comment (1Like1Prayer) [inline src/ts/Product.ts:67]: why let?

Comment (1Like1Prayer) [inline src/ts/Product.ts:67]: why cart1? do you have another cart in the system?

File: src/ts/Product.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    console.log("empty cart");
+    console.log(cart1);
```
Comment (1Like1Prayer) [inline src/ts/Product.ts:69]: you can write it in the same line - `console.log('empty cart', cart1)` 
either way, use template literal

File: src/ts/Product.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    public constructor(name: string, description: string, price: number, img: string, limit: number) {
+        this.name = name;
+        this.description = description;
+        this.price = price;
+        this.img = img;
+        this.limit = limit;
+    }
+
+    public updateProductAmount(limit: number) {
```
Comment (1Like1Prayer) [inline src/ts/Product.ts:16]: this should be a function in the cart class

File: src/ts/Product.ts (hunk 1 above)
Comment (1Like1Prayer) [inline src/ts/Product.ts:58]: products should be read from json and not created manually here

File: src/ts/Product.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    public removeProduct(product: Product) {
+        delete this.productsInTheCart[product.name];
+    }
+
+
+    public getTotalPrice(stock: Product[]) {
+        const arrayOfproductsInTheCart = Object.keys(this.productsInTheCart);
+        const totalPrice = arrayOfproductsInTheCart.reduce((total, item) => {
+            return total + (stock.find(product => product.name === item).price) * this.productsInTheCart[item]
```
Comment (1Like1Prayer) [inline src/ts/Product.ts:38]: use arrow function

File: src/ts/Product.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+(function main() {
+    const prod1 = {name: "Bamba", price: 5, limit: 10};
+    const prod2 = {name: "Bisli", price: 7, limit: 15};
+    const prod3 = {name: "Apropo", price: 4, limit: 20};
+    const stock = [prod1, prod2, prod3];
+
+    const cart = new Cart();
+    for (let newProduct in stock) {
```
Comment (1Like1Prayer) [inline src/ts/Product.ts:60]: don't use for loops. read about map

File: src/ts/Product.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        const arrayOfproductsInTheCart = Object.keys(this.productsInTheCart);
+        const totalPrice = arrayOfproductsInTheCart.reduce((total, item) => {
+            return total + (stock.find(product => product.name === item).price) * this.productsInTheCart[item]
+        }, 0);
+        console.log('total price: ', totalPrice);
+    }
+
+
+    public checkout(stock: Product[]) {
+        for (let productName in this.productsInTheCart) {
```
Comment (1Like1Prayer) [inline src/ts/Product.ts:45]: we try not to use for loops in js/ts. convert to array and use map

File: src/ts/Product.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    const cart = new Cart();
+    for (let newProduct in stock) {
+        cart.addProduct(stock[newProduct]);
+    }
+
+    console.log('Stock:',stock);
+    cart.updateProductAmount(prod2, 2);
+    cart.getTotalPrice(stock);
+    cart.removeProduct(prod1);
+    console.log("Cart after remove " + prod1.name, cart);
```
Comment (1Like1Prayer) [inline src/ts/Product.ts:68]: string literal

File: src/ts/Product.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    console.log('before checkout',stock);
```
Comment (1Like1Prayer) [inline src/ts/Product.ts:69]: print the cart, we need to see if the checkout works.

File: src/ts/Product.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    cart.checkout(stock);
+    console.log('after checkout',stock);
```
Comment (1Like1Prayer) [inline src/ts/Product.ts:71]: try to add a test of adding a product with limit 0 to see what happens

File: src/ts/Product.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+class Cart {
+    productsInTheCart: Record<string, number>
+
+    constructor() {
+        this.productsInTheCart = {};
+    }
+
+    public addProduct(product: Product) {
+        if (product.limit > 0)
```
[the remark this answers is not in the document]
Comment (DanielDavidyan) [reply src/ts/Product.ts:17]: Yes, it prints ״sold out" and does not add the product to the cart

**Pull request closed without merging**